### PR TITLE
Premio 20201203

### PR DESCRIPTION
**Pull request by eoliveros**
This appears to send the links. set it to 30 sec.

### CODE REVIEW

File: captain-definition [hunk 1]
```diff
@@ -8,6 +8,7 @@
                 "RUN apk add --no-cache make gcc libc-dev libxslt-dev postgresql-dev libffi-dev",
                 "RUN pip3 install --upgrade pip",
                 "RUN pip3 install -r requirements.txt",
+                "RUN python3 ./init_wallet_address.py",
```
Comment (djpnewton) [inline captain-definition]: this needs to be removed, it will break docker because that file no longer exists

File: requirements.txt [hunk 2]
```diff
@@ -1,7 +1,5 @@
-# greenlet version 0.4.14 has ABI change which causes crash on win/osx
```
Comment (djpnewton) [inline requirements.txt]: none of the changes to this file are required, also doubles up some dependancies

File: web.py [hunk 3]
```diff
@@ -19,6 +22,9 @@
 import base58
 import pywaves
 import pyblake2
+import qrcode
+import qrcode.image.svg
+from gevent_tasks import TaskManager, cron
```
Comment (djpnewton) [inline web.py]: this is imported but not used

File: web.py [hunk 4]
```diff
@@ -354,12 +377,19 @@ def start_greenlets():
             logger.info("starting WebGreenlet runloop...")
             self.runloop_greenlet.start()
 
+        def send_payment_greenlet():
+            while True:
+                gevent.spawn(process_proposals)
+                gevent.sleep(30)
+            self.runloop_greenlet.start()
```
Comment (djpnewton) [inline web.py]: this is never run (nor should be)

File: web.py [hunk 5]
```diff
@@ -354,12 +377,19 @@ def start_greenlets():
...
+
         # create greenlet
         self.runloop_greenlet = gevent.Greenlet(runloop)
         if self.exception_func:
             self.runloop_greenlet.link_exception(self.exception_func)
-        # check node/wallet and start greenlets
+        ## check node/wallet and start greenlets
         gevent.spawn(start_greenlets)
+        gevent.spawn(send_payment_greenlet)
```
Comment (djpnewton) [inline web.py]: remove this, add your greenlet to 'start_greenlets'

File: web.py [hunk 6]
```diff
@@ -354,12 +377,19 @@ def start_greenlets():
...
             self.runloop_greenlet.start()
 
+        def send_payment_greenlet():
+            while True:
+                gevent.spawn(process_proposals)
+                gevent.sleep(30)
+            self.runloop_greenlet.start()
+
         # create greenlet
         self.runloop_greenlet = gevent.Greenlet(runloop)
```
Comment (djpnewton) [inline web.py:485]: Do something similar to this:

`        self.process_proposals_greenlet = gevent.Greenlet(process_proposals_loop)`

File: web.py [hunk 7]
```diff
@@ -354,12 +377,19 @@ def start_greenlets():
...
         if self.exception_func:
             self.runloop_greenlet.link_exception(self.exception_func)
-        # check node/wallet and start greenlets
+        ## check node/wallet and start greenlets
         gevent.spawn(start_greenlets)
+        gevent.spawn(send_payment_greenlet)
 
     def stop(self):
         self.runloop_greenlet.kill()
```
Comment (djpnewton) [inline web.py:493]: then you can add:
`self.process_proposals_greenlet.kill()` and you will actually be able to stop the server

File: web.py [hunk 8]
```diff
@@ -165,41 +179,42 @@ def index():
 
 @app.route("/internal/process_proposals")
```
Comment (djpnewton) [inline web.py]: if the `with app.app_context():` line means that this function no longer works as a direct endpoint then this routing should be removed (and move the function out of the routes section), and remove the link to this endpoint from index.html

**Review by djpnewton — CHANGES_REQUESTED** [review]
Requires some changes in order to work properly (ie shutdown server with ctrl-c, install with docker)

Also some unneeded changes can be removed

File: captain-definition (hunk 1 above)
Comment (eoliveros) [reply captain-definition]: removed that line.

File: web.py (hunk 3 above)
Comment (eoliveros) [reply web.py]: removed gevent_tasks

File: web.py (hunk 4 above)
Comment (eoliveros) [reply web.py]: removed.

File: web.py (hunk 5 above)
Comment (eoliveros) [reply web.py]: removed this lines.

File: web.py (hunk 7 above)
Comment (eoliveros) [reply web.py:493]: added self.process_proposals_greenlet.kill()

File: requirements.txt (hunk 2 above)
Comment (eoliveros) [reply requirements.txt]: placed back what i deleted and removed the others and add qrcode

File: web.py (hunk 3 above)
Comment (eoliveros) [reply web.py]: removed urllib3

File: web.py (hunk 8 above)
Comment (eoliveros) [reply web.py]: removed the @app.route for "internal/process_proposals" and moved out of the route views.

File: requirements.txt [hunk 9]
```diff
@@ -1,4 +1,4 @@
...
+# greenlet version 1.4.14 has ABI change which causes crash on win/osx
```
Comment (djpnewton) [inline requirements.txt]: this comment change is incorrect

File: templates/claim_payment.html [hunk 10]
```diff
@@ -8,21 +8,24 @@ <h4>Expiry Date:</h4>
...
             <h4>Message:</h4>
             <p>{{payment.message}}</p>
-            <h4>Claim form:</h4>
+            <h4>Claim payment:</h4>
             <form method="post">
             {%if not recipient%}
-                <input type="text" name="recipient" placeholder="Wallet address"/>
+                <div class="card-body">
+                    <a href="{{ url }}">{{ qrcode_svg | safe }}</a><br/>
+                    <p>SCAN or CLICK the QRCode to claim your zap.</p><br/>
```
Comment (djpnewton) [inline templates/claim_payment.html]: change to `SCAN or CLICK the QR code to claim your {{config["ASSET_NAME"]}}.`

File: web.py [hunk 11]
```diff
@@ -348,11 +369,18 @@ def runloop():
...
+            while True:
+                gevent.spawn(process_proposals)
+                gevent.sleep(2)
+
         def start_greenlets():
             logger.info("checking wallet...")
             self.check_wallet()
             logger.info("starting WebGreenlet runloop...")
             self.runloop_greenlet.start()
+            self.process_proposals_greenlet = gevent.Greenlet(process_proposals_loop)
```
Comment (djpnewton) [inline web.py]: move this line to where the other greenlet is created

File: web.py [hunk 12]
```diff
@@ -348,11 +369,18 @@ def runloop():
...
+            self.process_proposals_greenlet.start()
```
Comment (djpnewton) [inline web.py:482]: you also need to add `self.process_proposals_greenlet.kill()` in the `stop()` function

**Review by djpnewton — CHANGES_REQUESTED** [review]
See notes for requested changes. Also need to pull from `premio_stage` branch in order to fix merge conflicts

File: requirements.txt (hunk 9 above)
Comment (eoliveros) [reply requirements.txt]: cahnged to 0.4.14

File: templates/claim_payment.html (hunk 10 above)
Comment (eoliveros) [reply templates/claim_payment.html]: changed to the one above.

File: web.py (hunk 11 above)
Comment (eoliveros) [reply web.py]: moved this line.

File: web.py (hunk 12 above)
Comment (eoliveros) [reply web.py:482]: created the greenlet.kill() for the new greenlet.

**Comment by djpnewton** [comment]
>Also need to pull from premio_stage branch in order to fix merge conflicts

fix merge conflicts please

File: web.py (hunk 6 above)
Comment (eoliveros) [reply web.py:485]: moved this to a function and its now being started under start_greenlets()

**Pull request merged**